src/tr2/game/gun/gun_misc.c

```diff
@@ -179,17 +196,28 @@ int32_t Gun_FireWeapon(
             view_pos.z + ((best_dist * g_MatrixPtr->_22) >> W2V_SHIFT);
         hit_pos.room_num =
             Room_GetIndexFromPos(hit_pos.pos.x, hit_pos.pos.y, hit_pos.pos.z);
-        const int16_t item_to_smash = LOS_CheckSmashable(&start, &hit_pos);
-        if (item_to_smash != NO_ITEM) {
-            Gun_SmashItem(item_to_smash, weapon_type);
-        }
+        Gun_SmashItems(start, hit_pos);
         Gun_HitTarget(
             target, &hit_pos,
             winfo->damage * (Game_IsBonusFlagSet(GBF_JAPANESE) ? 2 : 1));
         return 1;
     }
 }
 
+bool Gun_SmashItems(const GAME_VECTOR start, const GAME_VECTOR target)
+{
+    bool broken = false;
+    while (true) {
```

Comment:
In Ice Palace, shooting the bell causes an infinite loop here. This seems to resolve it, but perhaps there is a better way.

```diff
diff --git a/src/tr2/game/los.c b/src/tr2/game/los.c
index a47787d4a..04917dc0b 100644
--- a/src/tr2/game/los.c
+++ b/src/tr2/game/los.c
@@ -307,7 +307,8 @@ int32_t LOS_CheckSmashable(
 
     for (int16_t item_num = 0; item_num < Item_GetTotalCount(); item_num++) {
         const ITEM *const item = Item_Get(item_num);
-        if (item->status == IS_DEACTIVATED) {
+        if (item->status == IS_DEACTIVATED
+            || (item->object_id == O_BELL && item->status == IS_ACTIVE)) {
             continue;
         }
         if (!Object_IsType(item->object_id, g_SmashableObjects)) {
```

Reply:
Added a check in `Gun_SmashItems`